modin/pandas/dataframe.py

```diff
@@ -1421,27 +1462,28 @@ def merge(
                 "can not merge DataFrame with instance of type "
                 "{}".format(type(right))
             )
-        if left_index is False or right_index is False:
-            return self._create_or_update_from_compiler(
-                self._query_compiler.join(
-                    right._query_compiler,
-                    how=how,
-                    on=on,
-                    left_on=left_on,
-                    right_on=right_on,
-                    left_index=left_index,
-                    right_index=right_index,
-                    sort=sort,
-                    suffixes=suffixes,
-                    copy=copy,
-                    indicator=indicator,
-                    validate=validate,
-                )
-            )
-        if left_index and right_index:
-            return self.join(
-                right, how=how, lsuffix=suffixes[0], rsuffix=suffixes[1], sort=sort
+        result = self.__constructor__(
+            query_compiler=self._query_compiler.merge(
+                right._query_compiler,
+                how=how,
+                on=on,
+                left_on=left_on,
+                right_on=right_on,
+                left_index=left_index,
+                right_index=right_index,
+                sort=sort,
+                suffixes=suffixes,
+                copy=copy,
+                indicator=indicator,
+                validate=validate,
             )
+        )
+
+        if sort:
```

Comment:
I think this should be handled in the query compiler.

Reply:
Ok, I will think on this.

Reply:
I think we could leave this logic here until we move a logic with `sort_values` in to compiler layer.

Reply:
#1754 is opened to move that logic

Reply:
It's great! If you are not going to do something else regarding #1577 in that PR, we can merge it into master that I pull these changes to my PR.